### DISCUSSION TITLE
자료구조 - 배열/연결리스트 서희&재찬

### DISCUSSION THREAD

**Discussion by kjc6735** [Shortudy-10th/tech-for-developer#41] [질의 응답]
- 배열과 연결리스트의 특징에 대해서 설명해주세요.
- Static Array와 Dynamic Array의 차이점은 무엇일까요?
- Array는 Random Access Data Structure의 하나의 예가 될 수 있다고 합니다. Random Access란 무엇일까요?
- 연결리스트의 종류에 대해 설명해주세요.
- 배열과 연결리스트의 특성에 기반하여, 각 자료구조를 사용하는 것이 다른 자료구조를 사용하는 것보다 유리한 상황에 대해 설명해주세요.
- 연결리스트에서 요소를 추가하거나 삭제하는 과정을 설명해주세요.

**Comment by kjc6735** [comment]
### ❓️배열과 연결리스트의 특징에 대해서 설명해주세요.

<br>

### 💬 답변
<blockquote>

## Array란?

![](https://velog.velcdn.com/images/lire_eruel/post/07fb3f0b-d853-49ae-ab9d-0fbfeb05bb0d/image.png)

Array는 프로그래밍에서 일련의 요소들을 순차적으로 저장하는 데이터 구조입니다. 이러한 요소들은 같은 데이터 타입을 가지며, 각 요소는 배열 안에서 인덱스(index)라고 불리는 숫자로 접근할 수 있습니다. 인덱스는 보통 0부터 시작하여 배열의 크기보다 하나 작은 값을 가집니다.

예를 들어, 5개의 정수를 저장하는 배열은 다음과 같이 표현할 수 있습니다:

```
[10, 20, 30, 40, 50]
```

## Array 장점

### 요소에 빠르게 접근할 수 있다.

가장 기본적인 자료구조인 Array는 논리적 저장 순서와 물리적 저장 순서가 일치한다.
따라서 인덱스로 해당 원소에 Big-O(1)에 해당 원소로 접근할 수 있다. 즉 random access가 가능하다.

## Array 단점

### 삽입이나 삭제가 비효율적이다.

![](https://velog.velcdn.com/images/lire_eruel/post/024a2dbf-7a78-4b84-8f2d-1ee6b3792435/image.png)

삽입 또는 삭제 과정에서는 해당 원소에 접근하여 작업을 완료한 뒤(O(1)), 또 한 가지의 작업을 추가적으로 해줘야 하기 때문에, 시간이 더 걸린다.
만약 삭제라고 한다면, 중간에 원소를 삭제하고 그 뒤부터 요소들을 shift해줘야 하기 때문에 시간 복잡도는 O(n)이 된다.
삽입 또한 마찬가지 이다.

### 크기가 고정적이다.

Array는 생성할 때 크기를 지정하고 이후에 크기를 동적으로 변경할 수 없습니다. 만약 배열의 크기를 초과하여 데이터를 추가하려면, 새로운 배열을 생성하고 기존의 데이터를 복사해야 합니다. 이로 인해 삽입과 삭제가 비효율적이며, 크기를 자주 변경해야 하는 경우에는 다른 자료구조를 고려해야 합니다.

# Linked List

![](https://velog.velcdn.com/images/lire_eruel/post/3b43068b-aefa-4b9d-a712-05caebd911b0/image.png)

Linked List(연결 리스트)는 배열의 삽입과 삭제가 비효율적인 문제를 해결하기 위한 자료구조입니다. Linked List는 각 노드(Node)들이 데이터와 다음 노드를 가리키는 포인터(주소)로 구성됩니다. 각 노드들이 연결되어 있기 때문에 순차적으로 데이터를 저장하지 않고, 논리적으로 연결되어 있는 형태로 데이터를 관리합니다.

## Linked List 장점

### 삽입이나 삭제가 빠르다.

![](https://velog.velcdn.com/images/lire_eruel/post/2792a595-dd26-44f8-888e-872433a7d0b2/image.png)
각각의 원소들은 자기 자신 다음에 어떤 원소인지만을 기억하고 있다.
삽입이나 삭제를 할 때 해당 부분만 다른 값으로 변경하면 되기 때문에 삽입과 삭제를 O(1)만에 해결할 수 있다.

### 동적 크기

배열과 달리 동적으로 크기를 조정할 수 있습니다. 필요에 따라 노드를 추가하거나 제거할 수 있습니다.

## Linked List 단점

### 결국 삽입이나 삭제가 O(n)시간이 필요하다.

원하는 위치에 삽입을 하고자 하면 원하는 위치를 Search 과정에 있어서 첫번째 원소부터 다 확인해봐야 한다는 것이다. Array 와는 달리 논리적 저장 순서와 물리적 저장 순서가 일치하지 않기 때문이다. 이것은 일단 삽입하고 정렬하는 것과 마찬가지이다. 이 과정 때문에, 어떠한 원소를 삭제 또는 추가하고자 했을 때, 그 원소를 찾기 위해서 O(n)의 시간이 추가적으로 발생하게 된다.
</blockquote>

<br>

**Comment by kjc6735** [comment]
### ❓️Static Array와 Dynamic Array의 차이점은 무엇일까요?

<br>

### 💬 답변
<blockquote>
Static array는 고정된 배열의 크기를 가지고 있다. 그래서 크기를 늘리기 위해서는 배열을 새로 만들고 복사하는 방법으로 사용해야한다. 
Dynamic array는 ArrayList를 사용하여 구현된다. 크기를 늘릴 수 있으며, 일정 크기의 배열을 생성하고, 만약에 넘어간다면 자동으로 크기를 늘려준다. 
</blockquote>

<br>

**Reply by LireEruel** [reply]
그럼 Dynamic Array가 항상 좋은거 아닌가요? 언제 Static Array를 사용하나요?

**Reply by kjc6735** [reply]
사용해야 할 메모리의 크기를 알 때 사용하는 것이 좋다.

**Comment by kjc6735** [comment]
### ❓️Array는 Random Access Data Structure의 하나의 예가 될 수 있다고 합니다. Random Access란 무엇일까요?

<br>

### 💬 답변
<blockquote>
 
**RandomAccess** 
- 데이터 구조 내의 원소(element)들에 무작위로(임의로) 접근할 수 있는 능력
- 배열은 메모리 상에서 연속적으로 배치되어 있으며, 각 원소는 인덱스를 사용하여 접근할 수 있습니다. 즉, 배열(Array)과 같은 Random Access Data Structure는 인덱스를 사용하여 원소에 빠르게 접근할 수 있습니다.
</blockquote>

<br>

**Comment by kjc6735** [comment]
### ❓️연결리스트의 종류에 대해 설명해주세요.
<br>

### 💬 답변
<blockquote>
단일 연결 리스트 - 각 요소는 데이터와 다음 노드를 가리키는 링크로 구성된다.
각 노드가 다음 노드만 가리키므로 노드들이 한 방향으로 연결된다.

이중 연결 리스트
단일 연결 리스트는 단방향이였다면, 이중 연결 리스트는 양방향이다. 다음 노드와 이전 노드로 연결된다. 단일 연결 리스트 삭제, 삽입이 빠르지만 메모리 사용량이 더 많다.

</blockquote>

<br>

**Comment by kjc6735** [comment]
### ❓️배열과 연결리스트의 특성에 기반하여, 각 자료구조를 사용하는 것이 다른 자료구조를 사용하는 것보다 유리한 상황에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>
배열은 크기가 고정되어 있기 때문에 배열의 크기를 알고, 인덱스로 빠르게 접근하는 경우에 유리하다. 연결 리스트는 배열의 크기를 잘 모르거나 삽입,삭제가 빈번하게 일어나는 경우에 유리하다.
</blockquote>

<br>

**Comment by kjc6735** [comment]
### ❓️연결리스트에서 요소를 추가하거나 삭제하는 과정을 설명해주세요.
<br>

### 💬 답변
<blockquote>

![](https://velog.velcdn.com/images/lire_eruel/post/2792a595-dd26-44f8-888e-872433a7d0b2/image.png)
각각의 원소들은 자기 자신 다음에 어떤 원소인지만을 기억하고 있다.
삽입이나 삭제를 할 때 해당 부분만 다른 값으로 변경하면 되기 때문에 삽입과 삭제를 O(1)만에 해결할 수 있다.

1. 삭제할 노드의 before 요소의 next를 삭제 노드의 next로 연결한다. 
2. 삭제할 노드를 삭제한다.

</blockquote>

<br>
